File: flask_app/app/dao/address_book_dao.py
from app.models.address_book import AddressBook
from app.models.base import db

class AddressBookDAO:
    @staticmethod
    def create(data):
        address = AddressBook(**data)
        db.session.add(address)
        db.session.commit()
        return address

    @staticmethod
    def get_by_user(user_id):
        return AddressBook.query.filter_by(user_id=user_id).all()

    @staticmethod
    def get_by_id(address_id):
        return AddressBook.query.get(address_id)

    @staticmethod
    def update(address, updates):
        for key, value in updates.items():
            setattr(address, key, value)
        db.session.commit()
        return address

    @staticmethod
    def delete(address):
        db.session.delete(address)
        db.session.commit()

File: java_membership_api/src/main/java/com/example/membership_api/dao/AddressBookDAO.java
package com.example.membership_api.dao;

import java.util.List;

import org.springframework.stereotype.Repository;

import com.example.membership_api.model.AddressBook;

import jakarta.persistence.EntityManager;
import jakarta.persistence.NoResultException;
import jakarta.persistence.PersistenceContext;
import jakarta.transaction.Transactional;

@Repository
@Transactional
public class AddressBookDAO {

    @PersistenceContext
    private EntityManager em;

    public AddressBook create(AddressBook address) {
        em.persist(address);
        return address;
    }

    public List<AddressBook> getByUser(String userId) {
        return em.createQuery("FROM AddressBook WHERE userId = :userId", AddressBook.class)
                 .setParameter("userId", userId)
                 .getResultList();
    }

    public AddressBook getById(String addressId) {
        return em.find(AddressBook.class, addressId);
    }

    public AddressBook getByIdUserId(String addressId, String userId) {
        try {
            return em.createQuery(
                    "SELECT a FROM AddressBook a WHERE a.id = :addressId AND a.userId = :userId",
                    AddressBook.class
                )
                .setParameter("addressId", addressId)
                .setParameter("userId", userId)
                .getSingleResult();
        } catch (NoResultException e) {
            return null; 
        }
    }

    public AddressBook update(AddressBook existing, AddressBook updates) {
        existing.setName(updates.getName());
        existing.setAddress(updates.getAddress());
        existing.setPincode(updates.getPincode());
        existing.setCity(updates.getCity());
        existing.setState(updates.getState());
        existing.setCountry(updates.getCountry());
        existing.setIsDefault(updates.getIsDefault());
        return existing;
    }

    public void delete(AddressBook address) {
        em.remove(address);
    }
}
